Allow filtering of iframe src urls

ryan-verys opened the pull request:
Hey @boutell looks like I'm the motivated developer 😄 . 

This PR makes it so the user can pass allowable urls as a property to the sanitizeHtml function which will strip out any urls from an iframe tag's src attribute that are not included in the allowable urls property array. 

This is accomplished by just adding a new property to the options object passed to the function like so: 

```
clean = sanitizeHtml(dirty, {
    allowedTags: ['p', 'iframe', 'a', 'img', 'i'],
    allowedAttributes: {
        'iframe': ['src', 'href'],
        'a': ['src', 'href'],
        'img': ['src']
    },
    allowedUrls: ['youtube.com', 'vimeo.com']
})
```

The PR contains some default urls allowed (youtube.com and vimeo.com), I'm happy adjust these or remove them as you see fit. I can simply add a conditional for passing `all` as the only property if that is something you think would be helpful. 

2 passing tests written.

ryan-verys commented:
Hi @boutell I know it's the holidays but just wanted to see if you had a chance to peek at this PR and give some thoughts on it? Happy holidays!

boutell commented:
Hi Ryan,

I get what you're trying to do, but a top level option called "allowedUrls"
sounds like it applies to much, much more than iframes. Also you're locking
down domains, really, not URLs in general.



On Thu, Dec 21, 2017 at 2:37 PM, ryan-verys <notifications@github.com>
wrote:

> Hi @boutell <https://github.com/boutell> I know it's the holidays but
> just wanted to see if you had a chance to peek at this PR and give some
> thoughts on it? Happy holidays!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/punkave/sanitize-html/pull/192#issuecomment-353439080>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAB9fRl7aPsMI6gruBCWULJsHVtduwt0ks5tCrN1gaJpZM4RF3jK>
> .
>



-- 


*THOMAS BOUTELL, CHIEF SOFTWARE ARCHITECT*
P'UNK AVENUE | (215) 755-1330  |  punkave.com

ryan-verys commented:
Yeah I agree with you on the wording there. Are you suggesting a name change for the 'allowedUrls' option or nesting inside maybe the 'allowedAttributes' property? I'd want to avoid fidgeting with how the function processes content passed to it and cleanly append to it.

ryan-verys commented:
Ok I made another commit with wording changed to `allowedIframeDomains`. I think this will help clarify any confusion about what to pass to it. I still think it needs to be a top level property to avoid messing too much with the functionality of the module.

ryan-verys commented:
Added a a README commit that documents changes.

ryan-verys commented:
Ok that makes sense. I'll make those changes and push forward because it sounds like you are open to this feature being added as long as it's done right.

boutell commented:
Yep!

On Tue, Jan 2, 2018 at 1:00 PM, ryan-verys <notifications@github.com> wrote:

> Ok that makes sense. I'll make those changes and push forward because it
> sounds like you are open to this feature being added as long as it's done
> right.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/punkave/sanitize-html/pull/192#issuecomment-354832184>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAB9fVCNrirg1a2weVQ0MhHUg2jXLiNPks5tGm6zgaJpZM4RF3jK>
> .
>



-- 


*THOMAS BOUTELL, CHIEF SOFTWARE ARCHITECT*
P'UNK AVENUE | (215) 755-1330  |  punkave.com

ryan-verys commented:
The function will now parse the url and compare the hostname with the allowedIframeDomains. If there is no hostname, i.e. a url was stuffed into the iframe in some other way) the matching frame.attribs is deleted.

Added test to confirm and updated README.

boutell commented on src/index.js:
```diff
@@ -207,10 +208,26 @@ function sanitizeHtml(html, options, _recursing) {
+              try {
+                var parsed = url.parse(value)
+                var whitelistedDomains = []
+                each(parsed, function(val, prop) {
```
This each loop is redundant now, you are only interested in: `parsed.hostname`, so just check that property.

ryan-verys replied:
Removed.

boutell commented on src/index.js:
```diff
@@ -207,10 +208,26 @@ function sanitizeHtml(html, options, _recursing) {
+                var whitelistedDomains = []
+                each(parsed, function(val, prop) {
+                  if (prop === 'hostname') {
+                    whitelistedDomains = options.allowedIframeDomains.filter(function(domain) {
```
`find` is faster than `filter` here and less weird because you can use it directly in an `if` statement.

ryan-verys replied:
Changed.

boutell commented on src/index.js:
```diff
@@ -207,10 +208,26 @@ function sanitizeHtml(html, options, _recursing) {
+                each(parsed, function(val, prop) {
+                  if (prop === 'hostname') {
+                    whitelistedDomains = options.allowedIframeDomains.filter(function(domain) {
+                      return val.includes(domain)
```
You are still using `includes` so a name like "totallyfakeyoutube.com" would be accepted. You should be using a regular expression or otherwise matching either an exact match, or a match for a period followed by the domain name, at the end of the string only.

ryan-verys replied:
I opted to go for an exact match to the hostname. I think the user needs to pass explicitly which hostname they would like to have matched. 

But, in some instances (like we see at our organization) a composer or redactor will stuff a localized url into the src so it looks something like `//www.youtube.com/embed/RTrYY5ycajY`, with no `http`. So the url parser doesn't pick up the host name, it just stuffs the whole url into the href property and makes host and hostname `null`. So I added some additional functionality to pull that apart using `split` and checking for an exact match again.

boutell commented on src/index.js:
```diff
@@ -207,10 +208,26 @@ function sanitizeHtml(html, options, _recursing) {
+                  delete frame.attribs[a];
+                  return;
+                } else {
+                  value = url.format(parsed)
```
No need to rebuild the URL with `format` since you never modified `parsed`. You can use the original value.

ryan-verys replied:
Removed

boutell commented on test/test.js:
```diff
@@ -656,11 +656,20 @@ describe('sanitizeHtml', function() {
+  });
+  it('Should not allow urls that do not have proper hostname', function() {
+    assert.equal(
+      sanitizeHtml("<iframe src='//www.youtube.com/embed/c2IlcS7AHxM'></iframe>", {
```
Wait, isn't this the kind of URL that *should* work?

ryan-verys replied:
Yes, as mentioned above I wrote a check for this instance when the url parser doesn't grab the hostname. Added different test that has a mismatched hostname, which passes.

ryan-verys commented:
Thanks for bearing with me as I'm juggling this on top of a bunch of other stuff.

boutell commented on README.md:
````diff
@@ -294,6 +298,61 @@ sanitizeHtml(
+For example:
+
+```javascript
+clean = sanitizeHtml(<p><iframe src="https://www.youtube.com/embed/nykIhs12345"></iframe><p>, {
````
JavaScript string quotation marks are missing from your examples.

ryan-verys replied:
doh!

boutell commented on package.json:
```diff
@@ -31,6 +31,7 @@
     "lodash.mergewith": "^4.6.0",
     "postcss": "^6.0.14",
     "srcset": "^1.0.0",
+    "url": "^0.11.0",
```
url is a built-in node module and should not be an npm dependency. Make sure this is all being tested against the built in url module and not whatever this might be...

ryan-verys replied:
ah yes, made this change.

boutell commented on src/index.js:
```diff
@@ -206,7 +207,39 @@ function sanitizeHtml(html, options, _recursing) {
+                  return domain === parsed.hostname;
+                })
+                if (!whitelistedDomains) {
+                  //check the href to see if the hostname didn't get picked up, doesn't mean it's invalid
```
This seems to be about the issue with URLs that start with //. If so you should be fixing the URL by adding the https: before passing it to url.parse, less hacky that way.

ryan-verys replied:
Yeah i was trying to think of other ways the url might be presented but it seems that this is the one issue. Now I check for `http:` and `https:` and just `concat` the value of it if it doesn't exist and starts with `//`

boutell commented on src/index.js:
```diff
@@ -206,7 +207,39 @@ function sanitizeHtml(html, options, _recursing) {
+                if (!whitelistedDomains) {
+                  //check the href to see if the hostname didn't get picked up, doesn't mean it's invalid
+                  var noHost = options.allowedIframeDomains.find(function(domain) {
+                    return parsed.href.includes(domain)
```
This looks like it would match the domain name anywhere in the URL, which is insecure as discussed.

boutell replied:
OK, I see how the check further down addresses this, but this entire branch of the code can be eliminated by fixing the // URLs and letting them go through url.parse normally.

ryan-verys replied:
Yeah took this part out.

ryan-verys commented:
@boutell when you get a chance I made a new commit. ⬆️

boutell commented on src/index.js:
```diff
@@ -206,10 +207,32 @@ function sanitizeHtml(html, options, _recursing) {
-
+            if (name === 'iframe' && a === 'src') {
+              //Check if value contains proper hostname prefix
+              if (value.indexOf('http:') < 0 && 
```
The presence of http: or https: somewhere else in the string, i.e. not at the beginning, isn't relevant and shouldn't cause it to be rejected. For instance a query string parameter of the string might be another URL, etc. but we're not being asked to directly fetch that.

All you need to do here is check whether it starts with `//` or not. If it starts with `//` it by definition doesn't start with one of the others.

ryan-verys replied:
Agreed!

boutell commented on README.md:
```diff
@@ -115,7 +118,8 @@ selfClosing: [ 'img', 'br', 'hr', 'area', 'base', 'basefont', 'input', 'link', '
 allowedSchemesByTag: {},
-allowProtocolRelative: true
+allowProtocolRelative: true,
+allowedIframeDomains: ['youtube.com', 'vimeo.com']
```
This sometimes-incorrect example, and the other example below where you attempt to document the need to "also include a hostname," plus the fact that the url parser returns a property called "hostname" and that's what you are actually comparing to, together convince me that this option really needs to be called `allowedIframeHostnames`. Then it is clear what you must do (list every allowed hostname, not just the domain name, or it won't be accepted).

ryan-verys replied:
Agreed!

boutell commented:
Couple things:

1. Looks like the tests are still looking for allowedIframeDomains?

2. If the allowedIframeDomains option is not set *at all* (it is ===
undefined), it must allow everything through, rather than denying
everything, for bc.

Thanks!

On Thu, Jan 11, 2018 at 5:36 PM, ryan-verys <notifications@github.com>
wrote:

> *@ryan-verys* commented on this pull request.
> ------------------------------
>
> In src/index.js
> <https://github.com/punkave/sanitize-html/pull/192#discussion_r161095861>:
>
> > @@ -206,10 +207,32 @@ function sanitizeHtml(html, options, _recursing) {
>                  return;
>                }
>              }
> -
> +            if (name === 'iframe' && a === 'src') {
> +              //Check if value contains proper hostname prefix
> +              if (value.indexOf('http:') < 0 &&
>
> Agreed!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/punkave/sanitize-html/pull/192#discussion_r161095861>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAB9fWN1ffIifOzeTg4f_Vw5jl5enG7dks5tJozpgaJpZM4RF3jK>
> .
>



-- 


*THOMAS BOUTELL, CHIEF SOFTWARE ARCHITECT*
P'UNK AVENUE | (215) 755-1330  |  punkave.com

boutell commented:
(Excuse me, #2 should say allowedIframeHostnames of course, it's hard to
stop (: )

On Fri, Jan 12, 2018 at 9:28 AM, Tom Boutell <tom@punkave.com> wrote:

> Couple things:
>
> 1. Looks like the tests are still looking for allowedIframeDomains?
>
> 2. If the allowedIframeDomains option is not set *at all* (it is ===
> undefined), it must allow everything through, rather than denying
> everything, for bc.
>
> Thanks!
>
> On Thu, Jan 11, 2018 at 5:36 PM, ryan-verys <notifications@github.com>
> wrote:
>
>> *@ryan-verys* commented on this pull request.
>> ------------------------------
>>
>> In src/index.js
>> <https://github.com/punkave/sanitize-html/pull/192#discussion_r161095861>
>> :
>>
>> > @@ -206,10 +207,32 @@ function sanitizeHtml(html, options, _recursing) {
>>                  return;
>>                }
>>              }
>> -
>> +            if (name === 'iframe' && a === 'src') {
>> +              //Check if value contains proper hostname prefix
>> +              if (value.indexOf('http:') < 0 &&
>>
>> Agreed!
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/punkave/sanitize-html/pull/192#discussion_r161095861>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AAB9fWN1ffIifOzeTg4f_Vw5jl5enG7dks5tJozpgaJpZM4RF3jK>
>> .
>>
>
>
>
> --
>
>
> *THOMAS BOUTELL, CHIEF SOFTWARE ARCHITECT*
> P'UNK AVENUE | (215) 755-1330  |  punkave.com
>



-- 


*THOMAS BOUTELL, CHIEF SOFTWARE ARCHITECT*
P'UNK AVENUE | (215) 755-1330  |  punkave.com

ryan-verys commented:
1. Oops. Updated. 
2. Are you sure? As I understand it (and tested myself), the `xtend` module implemented at L73 inputs the default properties if no property is passed at all. I added defaults at L536. If the user passes an empty array that's intentionally to allow nothing through. As I mentioned, I'm happy to change the defaults or not have any but I think YouTube and Vimeo are pretty widely known video services.

boutell commented:
I missed that you had configured defaults. This is definitely not going to
work — it is a backwards incompatible change and this is not worth changing
the major version number over (: It will cause havoc for people already
allowing src for iframe based on reasonable expectations of how this module
behaves.

The option should not be set at all by default, and if it is not set at
all, then the feature should not operate.


On Mon, Jan 15, 2018 at 1:39 PM, ryan-verys <notifications@github.com>
wrote:

>
>    1. Oops. Updated.
>    2. Are you sure? As I understand it (and tested myself), the xtend
>    module implemented at L73 inputs the default properties if no property is
>    passed at all. I added defaults at L536. If the user passes an empty array
>    that's intentionally to allow nothing through. As I mentioned, I'm happy to
>    change the defaults or not have any but I think YouTube and Vimeo are
>    pretty widely known video services.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/punkave/sanitize-html/pull/192#issuecomment-357760063>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAB9ff0dWEUDRa45cSIUSlXWlzjYaNOMks5tK5tRgaJpZM4RF3jK>
> .
>



-- 


*THOMAS BOUTELL, CHIEF SOFTWARE ARCHITECT*
P'UNK AVENUE | (215) 755-1330  |  punkave.com

ryan-verys commented:
Oh yeah, no problem. This should do the trick. Thanks for clarifying 💯.

boutell commented:
Hmm, it's strange not to have any option for support for relative URLs (on the same site). But, that's a feature that may or may not have an audience. It can be added later if desired.

Thank you for your patient contribution!

boutell commented:
This was published to npm.

ryan-verys commented:
Happy to contribute if you find that it does have an audience!

boutell commented:
@ryan-verys would you please take a look at this new issue that came in about the docs for this feature? https://github.com/punkave/sanitize-html/issues/215
